src/main/kotlin/com/terraformation/backend/api/Gpx.kt

```diff
@@ -0,0 +1,78 @@
+package com.terraformation.backend.api
+
+import com.opencsv.CSVWriter
+import io.ktor.utils.io.charsets.name
+import java.io.ByteArrayOutputStream
+import java.io.OutputStream
+import java.nio.charset.StandardCharsets
+import javax.xml.stream.XMLOutputFactory
+import javax.xml.stream.XMLStreamWriter
+import org.springframework.http.ContentDisposition
+import org.springframework.http.HttpHeaders
+import org.springframework.http.HttpStatus
+import org.springframework.http.ResponseEntity
+
+/**
+ * Constructs an HTTP response in GPX format.
+ *
+ * @param filename Filename that will be used by default if the user saves the GPX file. If null,
+ *   response will not include a filename.
+ * @param columnNames Contents of the first row of the CSV; this should be a list of the names of
+ *   the columns, typically in human-readable form.
+ * @param writeRows Callback function that writes the data rows to the CSV stream. This should call
+ *   [CSVWriter.writeNext] for each data row.
+ */
+fun gpxResponse(
+    filename: String?,
+    waypoints: List<GpxWaypoint>,
+): ResponseEntity<ByteArray> {
+  val byteArrayOutputStream = ByteArrayOutputStream()
+
+  GpxWriter(byteArrayOutputStream, waypoints).write()
+
+  val value = byteArrayOutputStream.toByteArray()
+  val headers = HttpHeaders()
+  headers.contentLength = value.size.toLong()
+  headers["Content-type"] = "application/gpx+xml;charset=UTF-8"
+
+  if (filename != null) {
+    headers.contentDisposition = ContentDisposition.attachment().filename(filename).build()
+  }
+
+  return ResponseEntity(value, headers, HttpStatus.OK)
+}
+
+// TODO: Add track, or routes data types if applicable
+// https://www.topografix.com/GPX/1/1/gpx.xsd (wptType)
+data class GpxWaypoint(val latitude: Double, val longitude: Double, val name: String?)
+
+class GpxWriter(output: OutputStream, private val waypoints: List<GpxWaypoint>) {
+  private val writer: XMLStreamWriter =
+      XMLOutputFactory.newInstance().createXMLStreamWriter(output, StandardCharsets.UTF_8.name)
+
+  private fun writeWaypoint(waypoint: GpxWaypoint) {
+    writer.writeStartElement("wpt")
+    writer.writeAttribute("lat", "${waypoint.latitude}")
+    writer.writeAttribute("lon", "${waypoint.longitude}")
```

Comment:
Does the GPX format allow all the string representations of floating-point numbers the JVM can produce? I'm thinking of stuff like a site near the equator ending up with a latitude value of `5.0E-9` using the default string rendering for `Double`s.

If not, we might need to use a `NumberFormat` to get precise control over number of decimal places and such.

Reply:
I opted to use BigDecimal set scale, because we don't actually want locale formatting.